crates/bevy_ecs/src/schedule/stage.rs

```diff
@@ -199,51 +227,102 @@ impl SystemStage {
-                match insertion_point {
-                    InsertionPoint::AtStart => {
-                        let index = self.exclusive_at_start.len();
-                        self.uninitialized_at_start.push(index);
-                        self.exclusive_at_start.push(container);
-                    }
-                    InsertionPoint::BeforeCommands => {
-                        let index = self.exclusive_before_commands.len();
-                        self.uninitialized_before_commands.push(index);
-                        self.exclusive_before_commands.push(container);
-                    }
-                    InsertionPoint::AtEnd => {
-                        let index = self.exclusive_at_end.len();
-                        self.uninitialized_at_end.push(index);
-                        self.exclusive_at_end.push(container);
-                    }
-                }
-            }
-            SystemDescriptor::Parallel(descriptor) => {
-                self.uninitialized_parallel.push(self.parallel.len());
-                self.parallel
-                    .push(ParallelSystemContainer::from_descriptor(descriptor, set));
-            }
+        let (run_criteria, mut systems) = system_set.bake();
+        let set_run_criteria_label =
+            run_criteria.map(|criteria| self.process_run_criteria(criteria));
+        for system in systems.drain(..) {
+            self.add_system_inner(system, set_run_criteria_label.clone());
         }
         self
     }
 
+    pub fn with_stage_run_criteria<S: System<In = (), Out = ShouldRun>>(
+        mut self,
+        system: S,
+    ) -> Self {
+        self.set_stage_run_criteria(system);
+        self
+    }
+
+    pub fn set_stage_run_criteria<S: System<In = (), Out = ShouldRun>>(
+        &mut self,
+        system: S,
+    ) -> &mut Self {
+        self.stage_run_criteria.set(Box::new(system));
+        self
+    }
+
+    pub fn with_run_criteria(mut self, run_criteria: RunCriteriaDescriptor) -> Self {
+        self.add_run_criteria(run_criteria);
+        self
+    }
+
+    pub fn add_run_criteria(&mut self, run_criteria: RunCriteriaDescriptor) -> &mut Self {
+        self.process_run_criteria(RunCriteriaDescriptorOrLabel::Descriptor(run_criteria));
+        self
+    }
+
+    fn process_run_criteria(
+        &mut self,
+        run_criteria: RunCriteriaDescriptorOrLabel,
+    ) -> BoxedRunCriteriaLabel {
+        let mut descriptor = match run_criteria {
+            RunCriteriaDescriptorOrLabel::Label(label) => return label,
+            RunCriteriaDescriptorOrLabel::Descriptor(descriptor) => descriptor,
+        };
+        let index = self.run_criteria.len();
+        self.uninitialized_run_criteria
+            .push((index, descriptor.duplicate_label_strategy));
+        let label = descriptor
+            .label
+            .take()
+            .unwrap_or_else(|| Box::new(RunCriteriaIndexLabel(index)));
```

Comment:
Do we really need RunCriteriaContainer to have a label? It seems like we could both remove the RunCriteriaIndexLabel type and just ignore `Option::<Label>::None` when deduping. That would also cut down on the amount of hashing / memory used.

Reply:
Right now, it's for this case:
```rust
SystemSet::new()
    .with_run_criteria(my_run_criteria.system())
    .with_system(stuff_1.system())
    .with_system(stuff_2.system())
```
This defines a new criteria once, and uses it for both systems. I guess it could be reworked to only do that for sets, but: we will need some stable way of referring to criteria when we implement runtime editing of stages. (It won't be sufficient as written, likely will need to be a sort of a generational index, or UUID.)

Reply:
I've realized that this issue runs pretty deep / this isn't a superficial changed. I'm wrapping up a refactor that I think both increases clarity and removes the need to throw "index labels" on criterias without labels.

Reply:
Just pushed the changes. Whether or not this "increases clarity" is up for debate. But it feels less like a "hack" than RunCriteriaIndexLabel.